qiskit_pkg/setup.py

```diff
@@ -0,0 +1,73 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2018.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# This file is the setup.py file for the qiskit package. Because python
+# packaging doesn't offer a mechanism to have qiskit supersede qiskit-terra
+# and cleanly upgrade from one to the other, there needs to be a separate
+# package always needs
+
+import json
+import os
+
+from setuptools import setup
+
+README_PATH = os.path.join(os.path.abspath(os.path.dirname(__file__)), "README.md")
+with open(README_PATH) as readme_file:
+    README = readme_file.read()
+
+EXTRAS_PATH = os.path.join(os.path.abspath(os.path.dirname(__file__)), "extras.json")
+with open(EXTRAS_PATH) as fd:
+    extras_dict = json.load(fd)
+
+extras_dict["all"] = [pkg for requirements in extras_dict.values() for pkg in requirements]
+
+requirements = ["qiskit-terra==0.45.0"]
+
+setup(
+    name="qiskit",
```

Comment:
Would it be possible / better to publish our wheels as `qiskit`, flip the dependency order, and have `qiskit-terra` become the wrapper package.  In other words, swap the name of what's currently `qiskit-terra` to become `qiskit`, and have this "new" package be `qiskit-terra` which just contains the dependency on `qiskit`?

As far as I could think, it'll satisfy all the upgrade paths correctly, and it also has the pleasant properties:
- the preferred install, `pip install qiskit`, installs directly (and the name Terra doesn't appear any more)
- our preferred install path links directly to the wheels, so going to the PyPI page is less surprising
- similarly, the PyPI page for qiskit-terra will much more clearly show that it's been superseded (and we could even change the README there to simply say that, rather than the symlink trick here because the _main_ package has remained Terra)
- we can drop `qiskit-terra` in the future, say at 1.0 or 2.0, and `pip install qiskit` will still work correctly
- we effectively gain another ~10GB storage on PyPI, since we've barely used any of our space on the metapackage
- it matches the directory structure of this repo more naturally

Reply:
It's the lack of uninstall of the old `qiskit-terra` package that's potentially the problem, right?

I wonder if we might actually just be better doing it anyway and saying "you might want to manually uninstall `qiskit-terra` when upgrading"?

Reply:
I really wish setuptools had a supersedes option that let us say `qiskit` supersedes the qiskit-terra package so it would correctly handle the upgrade path. I guess it's enough of an edge case that it doesn't come up very often, but for real package managers they typically offer this in the metadata because it does happen.

Reply:
Here are my thoughts (as someone who will not hear anywhere near as many user complaints as either of you :slightly_smiling_face:):

I don't love the dependency reversal because of the scenario @mtreinish mentioned of `pip install -U qiskit` overwriting an old `qiskit-terra`'s contents without removing it and also because it will break all the downstream projects that depend now explicitly on `qiskit-terra` without an upper version bound. What about this progression:

1. Add a check to `qiskit/__init__.py` that warns if `qiskit` is not installed but leave all the code in `qiskit-terra` and keep `qiskit` as a package depending only on a specific version of `qiskit-terra`.
2. After enough time for the community to have switched over to depending on and install `qiskit` rather than `qiskit-terra` (a year?), make a release with the code moved into the `qiskit` package with `qiskit-terra` still as a dependency but now empty.
3. After enough time again (another year?) that most users should have a release with the code in the `qiskit` package, drop the terra dependency.

The goal is to make sure that everyone is using `qiskit` before shifting the code over, so that the upgrade to an empty terra does not leave someone with no files. It requires a lot of patience though.

Personally, I think it would be better to get rid of terra if not too painful because it will be a point of minor confusion, mostly as a distraction but also some potential confusion since pip does allow some ways of installing packages without dependencies which could lead a user to end up with `qiskit` and `qiskit-terra` with different version numbers.

Reply:
Will: I pretty much agree with everything you said, the reason I was pushing for a more compressed timeline was that I think we'll get rather more acceptance from users between the 0.x -> 1.0 transition that they might need to `pip uninstall x y z; pip install x` to upgrade than even the 1.0 -> 2.0 transition, even if there are warnings.  I'm also in favour of getting rid of `qiskit-terra`, because I think it'll _hopefully_ be a minor confusion in most cases, but in cases where it does matter, we're going to have to talk users through more pain that it'll cause here, and it'll be a permanent albatross for us if we don't cut it when we deem we "have the chance".

That said, for this immediate PR, we're going to punt the decision a little bit, since the current form is what the metapackage currently does, and we wouldn't do this until at least the 1.0 release, if we do it at all.